### PR TITLE
Added option to change strategy for publishing events

### PR DESCRIPTION
**Pull request by Nyholm**
We are currently pushing all events on the message queue. If you got the following scenario: 
- Same application producing and consuming messages
- You have a lot of events fired 
- Only a few asynchronous events

... you will have a big overhead on your server. The background process starts the application to find out that there is nothing to do with the event. 

This PR makes sure that we only publish events that got an async subscriber. This feature is implemented as an opt-in feature. 

``` yaml
simple_bus_asynchronous:
  events:
    strategy: 'predefined' # default: 'always'
```

If this is a feature that you like Im happy to finish this PR with tests and docs. 
### TODO
- [x] Add tests
- [x] Add docs
### Related

Related to: https://github.com/SimpleBus/Asynchronous/pull/7

### CODE REVIEW

File: src/DependencyInjection/Compiler/CollectAsyncEventNames.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace SimpleBus\AsynchronousBundle\DependencyInjection\Compiler;
+
+use SimpleBus\SymfonyBridge\DependencyInjection\Compiler\CollectServices;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+class CollectAsyncEventNames implements CompilerPassInterface
```
Comment (ruudk) [inline src/DependencyInjection/Compiler/CollectAsyncEventNames.php]: Maybe `CollectAsyncronousEventNames` would be better?

File: src/Resources/config/asynchronous_events.yml [hunk 2]
```diff
@@ -13,9 +13,14 @@ services:
         public: false
         arguments:
             - '@simple_bus.asynchronous.event_publisher'
-        tags:
-            # insert before the middleware that actually notifies a message subscriber of the message
-            - { name: event_bus_middleware, priority: 0 }
+
+    simple_bus.asynchronous.published_predefined_messages_middleware:
```
Comment (ruudk) [inline src/Resources/config/asynchronous_events.yml]: Should be `publishes` instead of `published`

**Comment by ruudk** [comment]
I like this approach. Our application also has the same problem. A couple of events need to be handled async, not all. 

What do you think @matthiasnoback ?

File: src/DependencyInjection/Compiler/CollectAsyncEventNames.php (hunk 1 above)
Comment (Nyholm) [inline src/DependencyInjection/Compiler/CollectAsyncEventNames.php]: I had a hard time naming this class. Any suggestions are welcome =)

I'll change.

File: src/Resources/config/asynchronous_events.yml (hunk 2 above)
Comment (Nyholm) [inline src/Resources/config/asynchronous_events.yml]: Thank you

File: src/DependencyInjection/Compiler/CollectAsyncEventNames.php (hunk 1 above)
Comment (ruudk) [inline src/DependencyInjection/Compiler/CollectAsyncEventNames.php]: Maybe it should be `CollectAsyncronousEvents`

It collects all events that are handled asynchronous.

**Comment by cmodijk** [comment]
We have the same problem this is the reason why we still use our own implementation. I was working on a similar pull request.

**Comment by matthiasnoback** [comment]
Yes, looks great! As mentioned, having some tests would be nice :)

**Comment by Nyholm** [comment]
Cool. Im happy to see that this PR was needed. I'll make sure to finish these PRs tonight.

**Comment by Nyholm** [comment]
The PRs are updated, (no longer any WIP) and they are ready to be reviewed.

**Comment by ruudk** [comment]
@matthiasnoback The coveralls comments are really annoying :see_no_evil:

**Comment by Nyholm** [comment]
@ruudk #7

**Comment by Nyholm** [comment]
I've been running this in production for a day now. It really helped me to reduce the load on the background threads.

**Comment by matthiasnoback** [comment]
Hi @Nyholm. I released a new version of the Asynchronous library. I guess the only thing that's needed before we can release a new version of this bundle as well is to require at least v2.1.0 of the library. Could you fix that please?

**Comment by Nyholm** [comment]
Of course, thank you.

File: src/DependencyInjection/SimpleBusAsynchronousExtension.php [hunk 3]
```diff
@@ -53,12 +53,7 @@ private function requireBundle($bundleName, ContainerBuilder $container)
     {
         $enabledBundles = $container->getParameter('kernel.bundles');
         if (!isset($enabledBundles[$bundleName])) {
-            throw new \LogicException(
-                sprintf(
-                    'You need to enable "%s" as well',
-                    $bundleName
-                )
-            );
+            throw new \LogicException(sprintf('You need to enable "%s" as well', $bundleName));
```
Comment (matthiasnoback) [inline src/DependencyInjection/SimpleBusAsynchronousExtension.php:56]: Hehe ;)

**Pull request merged**